### PR TITLE
Structural/adding kinematic hardening

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding the kinematic plasticity with 3 different kinematic hardening allowed:

- Linear
- Amstrong-Frederick
- Araujo-Voyiadjis

The structure is the same as the isotropic plasticity with several methods added.
This implementation should allow to use kinematic+isotropic hardening with the isotropic laws that we already have.

Everything is implemented, I only have to perform some testing and do the tests.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 1]
```diff
@@ -0,0 +1,1044 @@
...
+     * @param rPlasticDissipation The internal variable of energy dissipation due to plasticity
+     * @param rPlasticStrainIncrement The increment of plastic strain of this time step
+     * @param rConstitutiveMatrix The elastic constitutive matrix
+     * @param rPlasticStrain The elastic constitutive matrix
+     * @param rValues Parameters of the constitutive law
+     * @param CharacteristicLength The equivalent length of the FE
+     * @param rBackStressVector The so-called BackStressVector used for the kinematic hardening
+     * @param rPreviousStressVector The previous converged stress vector
+     */
+    static void IntegrateStressVector(
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:164]: Mare meua, even more arguments. We should create structs to compact this

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 2]
```diff
@@ -0,0 +1,1044 @@
...
+        )
+    {
+        bool is_converged = false;
+        IndexType iteration = 0, max_iter = 100;
+        array_1d<double, VoigtSize> delta_sigma;
+        double plastic_consistency_factor_increment, F;
+
+        // Backward Euler
+        while (is_converged == false && iteration <= max_iter) {
+            F = rUniaxialStress - rThreshold;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h]: Please, look how I do this in the finite strain computation in order to compute just once

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 3]
```diff
@@ -0,0 +1,1044 @@
...
+
+            case KinematicHardeningType::AraujoVoyiadjisKinematicHardening:
+                KRATOS_ERROR_IF(kinematic_parameters.size() != 3) << "Kinematic Parameters not defined..." << std::endl;
+                dot_product_dp = 0.0;
+                for (IndexType i = 0; i < rPlasticStrainIncrement.size(); ++i) {
+                    dot_product_dp += rPlasticStrainIncrement[i] * rPlasticStrainIncrement[i];
+                }
+                pDot = std::sqrt(2.0 / 3.0 * dot_product_dp);
+                denominator = 1.0 + (kinematic_parameters[1] * pDot);
+                if (pDot > tolerance) {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:332]: Just to be sure, this is right or there is a std::abs missing?

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h [hunk 4]
```diff
@@ -0,0 +1,1044 @@
...
+        KRATOS_ERROR_IF(Gt3 < 0.0) << "Fracture energy too low in CurveFittingHardening of plasticity..."  << std::endl;
+
+        // Compute segment threshold
+        const double segment_threshold = (Gt2 + Gt1) / fracture_energy;
+
+        if (PlasticDissipation <= segment_threshold) {
+            const double Eps = EquivalentPlasticStrain;
+
+            if (EquivalentPlasticStrain < plastic_strain_indicator_1) { // Polinomial region
+                double S_Ep = curve_fitting_parameters[0];
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:728]: In general, not just here, or more descriptive name to the variables or comments to know what is each variable

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp [hunk 5]
```diff
@@ -118,7 +118,6 @@ void GenericSmallStrainIsotropicDamage<TConstLawIntegratorType>::CalculateMateri
 
             if (r_constitutive_law_options.Is(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR)) {
                 noalias(r_tangent_tensor) = (1.0 - damage) * r_constitutive_matrix;
-                TConstLawIntegratorType::YieldSurfaceType::CalculateEquivalentStress(auxiliar_integrated_stress_vector, r_strain_vector, uniaxial_stress, rValues);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp:121]: Can this be corrected in finite strain case?

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+{
+    if (rThisVariable == INTEGRATED_STRESS_TENSOR) {
+        //1.-Compute total deformation gradient
+        const Matrix& DeformationGradientF = rParameterValues.GetDeformationGradientF();
+        //2.-Right Cauchy-Green tensor C
+        const Matrix right_cauchy_green = prod(trans(DeformationGradientF), DeformationGradientF);
+        Vector strain_vector = ZeroVector(6);
+
+        //E= 0.5*(FT*F-1) or E = 0.5*(C-1)
+        strain_vector[0] = 0.5 * (right_cauchy_green(0, 0) - 1.00);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp]: Look in the constituve_law_utilities. There is an implementation of this

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp [hunk 7]
```diff
@@ -194,6 +194,8 @@ KRATOS_CREATE_VARIABLE(Matrix, PLASTIC_STRAIN_TENSOR)
 KRATOS_CREATE_VARIABLE(Vector, CURVE_FITTING_PARAMETERS)
 KRATOS_CREATE_VARIABLE(Vector, PLASTIC_STRAIN_INDICATORS)
 KRATOS_CREATE_VARIABLE(double, EQUIVALENT_PLASTIC_STRAIN)
+KRATOS_CREATE_VARIABLE(Vector, KINEMATIC_PLASTICITY_PARAMETERS)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:197]: Just one thing, be sure to define what each parameter means

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minos aesthetic comments, I you ask me I will approve, but better to take a look and clarify some variable definition among other things.

I will add the finite strain version after this merge

We should separate the laws from the others declarations or the main application file will become huge

**Comment by loumalouomega** [comment]
@AlejandroCornejo look at the warnings from Travis, there are important

**Comment by RiccardoRossi** [comment]
@AlejandroCornejo why is this a "FastPR"? 

In priciple FastPRs are just for cosmetic changes...this is definitely not the case

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:332]: this variable is always positive by force but thank you hehe

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h:728]: sometimes it is difficult but we can come up with some names...

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp (hunk 5 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp:121]: the damage is not in finite strain doesn't it?

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp (hunk 6 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp]: I'll remove thx

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp (hunk 7 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.cpp:197]: yes! there are 3 different models in kinematic plasticity, one model requires 1 param, the 2nd requires 2 and the more complex, 3. I'll describe it in the wiki but there are checks that gives an error if the size of the param vector is not the minimun required..

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_constitutive_law_integrator_kinematic_plasticity.h]: done ^^

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp (hunk 6 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_kinematic_plasticity.cpp]: solved

File: applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp (hunk 5 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_isotropic_damage.cpp:121]: True

**Review by loumalouomega — APPROVED** [review]

**Comment by loumalouomega** [comment]
Looks now Ok for me, let's proceed

**Comment by philbucher** [comment]
@loumalouomega @AlejandroCornejo this PR introduced a very large warning => you can see it in the [travis-log](https://travis-ci.org/KratosMultiphysics/Kratos/builds/450822186?utm_source=github_status&utm_medium=notification)

can you please fix it?
it is very hard to see other warnings

**Comment by loumalouomega** [comment]
@AlejandroCornejo can you do it?

**Comment by AlejandroCornejo** [comment]
okk sorry I didn't remember to look that

**Comment by AlejandroCornejo** [comment]
I'll solve this with a PR this afternoon

**Comment by philbucher** [comment]
thanks!

**Pull request merged**